parquet/src/main/java/org/apache/iceberg/parquet/ParquetReader.java

```diff
@@ -130,13 +139,23 @@ private void advance() {
 
       PageReadStore pages;
       try {
-        pages = reader.readNextRowGroup();
+        // Because of the issue of PARQUET-1901, we cannot blindly call readNextFilteredRowGroup()
+        if (hasRecordFilter) {
+          pages = reader.readNextFilteredRowGroup();
+        } else {
+          pages = reader.readNextRowGroup();
+        }
       } catch (IOException e) {
         throw new RuntimeIOException(e);
       }
 
+      long blockRowCount = blocks.get(nextRowGroup).getRowCount();
+      Preconditions.checkState(blockRowCount >= pages.getRowCount(),
+              "Number of values in the block, %s, does not great or equal number of values after filtering, %s",
+              blockRowCount, pages.getRowCount());
       long rowPosition = rowGroupsStartRowPos[nextRowGroup];
```

Comment:
I am not sure if the `nextRowGroup` counter and consequently `rowPosition` will be accurate in case of skipped row groups. Can you add a test to `TestSparkParquetReadMetadataColumns` which would invoke page skipping?

Reply:
After reading more about Parquet and page skipping, I don't think my comment above holds true in the general case. Seems like `nextRowGroup` counter should still accurately depict the current row group being processed. The only edge case that I see is in https://github.com/apache/parquet-mr/blob/fe9ab5d22c602776f44b027ad3cc05609eddba54/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java#L967 where a whole row group may be skipped if the column index allows all the pages in the row group to be skipped. In such a case, `readNextFilteredRowGroup` will automatically advance to the next block without our knowledge and hence without the counter being updated. This will be an issue for all further row group skipping and row postition calculations in Iceberg's ParquetReader. However, I am not sure if there can be such a case where Iceberg's row group skipping is not able to skip the row group, but Parquet's page skipping ends put skipping all page and consequently the whole row group.

Reply:
This edge case is probably possible with a carefully crafted file. e.g. consider the min/max values of single column
Row Group: `[1, 1000]`
Pages: `[[1, 10], [900, 100]]`
And expression `column = 500` will match the row group but will not match any of the pages resulting in the next row group being read automatically without the `nextRowGroup` counter being updated

Reply:
This is such a critical comment! That is absolutely possible and we need to handle that case. I looked at the code again and found even the existing code, we are missing a condition check 'pages == null'. It seems valid that readNextRowGroup() can return null https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java#L899. In that case, the caller needs to handle while it doesn't.

Reply:
@shardulm94, after looking into the code deeper, I think there could be two solutions here. 1) Change Parquet to let readNextFilteredGroup() return not only the filtered pages but also the record counts it skipped. 2) In the ReadConf constructor, when deciding 'shouldRead', we add ColumnIndex filter. This will ensure if a row group is called by readNextFilteredGroup(), there will always be at least 1 page returned and no implicitly advance internally without Iceberg knowing it. Here is the place for determining the 'shouldRead' boolean variable. https://github.com/apache/iceberg/blob/master/parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java#L106

For option 1, we will need to wait for the new release of Parquet. I am not sure releasing 1.10.2 is an option for Parquet. If not, we have to wait for Parquet 1.12.0. For option 2, it requires a pretty significant change. And we have to make sure the implementation here is consistent with Parquet' ColumnIndex filter implementation. 

Please share your thought and we can discuss.

Reply:
@shardulm94, I talked to the Parquet community(PARQUET-1927) and it seems we don't need Parquet change anymore. What we can do is to use ParquetFileReader constructor's filter(page level stats, row group stats, dictionray and future bloomfilter). All the following calls like reader.getRowGroups() and reader.getFilteredRecordCount() in ReadConf are all filtered values. It also simplifies the ReadConf constructor and ParquetReader/VectorizedParquetReader.

Reply:
@shangxinli After going through the code, I don't think this solves the problem. `ParquetFileReader.getRowGroups` does not apply column index filters (only stats and dict). So there is still the case where `getRowGroups` will return a block which will be skipped by `readNextFilteredGroup()`. I think the current test case does not accurately capture this.

Overall I don't think page skipping plays nicely with Iceberg's requirement to accurately fetch row numbers for row level deletes. e.g Iceberg currently stores the row number of the start of each row group, but in case of page skipping, some pages in the middle of the row group may be skipped and we would have no way to figure out the accurate row numbers. So we may have to disable page skipping when we want to fetch row numbers.

cc: @rdblue 
Having thought about this issue deeply in the Parquet context, I have doubts whether our ORC implementation for row numbers also suffers from the same issue. I will have to revisit this for ORC.